### PR TITLE
WIP: Refactoring yunohost format arguments parsing

### PR DESCRIPTION
**Pull request by Psycojoker**
## The problem

The current argument parsing code is rigged with quite some problems:

- everything is mixed together
- everything is parsed then everything is asked at once, this is very confusing and hard to extend
- all different kinds or arguments are parsed in the same place mixing all the things
- when "ask" is not present behavior is super weird
- code is hard to understand
- parsing individual arguments is closely tide to walking the data structure
- etc....

## Solution

After having written test for everything (oh god this was so freaking boring), refactor everything.

The approach is to factorize everything is a main generic parsing argument class then inherit from it for all arguments.

This will allow to extends this argument system and had new kind of arguments very easily.

Also everything is easier to read now and everything is nicely put in its small place.

And since we have tests we can refactor it again is needed in the future.

## PR Status

Tests pass and this is a very isolated part and now well tested part of the system so this shouldn't change anything at all for the users.

The only BREAKING change is that it will raises now if one app dev adds a default value to a password type of argument because I never want that.

## How to test

Launch tests or install an application.

## Validation

- [ ] Principle agreement 0/2 : 
- [ ] Quick review 0/1 : 
- [ ] Simple test 0/1 : 
- [ ] Deep review 0/1 :

### CODE REVIEW

File: locales/en.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "app_already_up_to_date": "{app:s} is already up-to-date",
     "app_argument_choice_invalid": "Use one of these choices '{choices:s}' for the argument '{name:s}'",
     "app_argument_invalid": "Pick a valid value for the argument '{name:s}': {error:s}",
```
Comment (kay0u) [inline locales/en.json:15]: Please, remove this keys:

- `app_argument_invalid`
- `domains_available`
- `pattern_password_app`
- `users_available`

Comment (Psycojoker) [reply locales/en.json:15]: I'm going to use them actually to make better error messages, just haven't found the energy yet.

Comment (Psycojoker) [reply locales/en.json:15]: Normally done

**Comment by Psycojoker** [comment]
Added some missing stuff for password and custom errors messages.

Should be way closer to being mergeable.

File: src/yunohost/app.py [hunk 2]
```diff
@@ -2525,128 +2736,14 @@ def _parse_args_in_yunohost_format(user_answers, argument_questions):
...
-                    msignals.display(m18n.n('users_available'))
-                    users = user_list()['users']
-                    for user in users.keys():
-                        msignals.display("- {}".format(user))
-
-                    root_mail = "root@%s" % _get_maindomain()
-                    for user in users.keys():
-                        if root_mail in user_info(user)["mail-aliases"]:
-                            question_default = user
-                            break
```
Comment (alexAubin) [inline src/yunohost/app.py:2572]: That stuff about default answer for user-type question disappeared during conflict/merge, looking at the new structure it should be pretty straightforward to reintegrate but busy right now with other reviews

**Review by alexAubin — CHANGES_REQUESTED** [review]

File: src/yunohost/app.py [hunk 3]
```diff
@@ -2514,6 +2514,217 @@ def _parse_args_for_action(action, args={}):
...
+
+
+class UserArgumentParser(YunoHostArgumentFormatParser):
+    argument_type = "user"
+
+    def parse_question(self, question, user_answers):
+        from yunohost.user import user_list
+
+        question = super(UserArgumentParser, self).parse_question(question, user_answers)
+        question.choices = user_list()["users"]
```
Comment (kay0u) [inline src/yunohost/app.py:2685]:
```suggestion
        question.choices = user_list()["users"]
        if question.default is None:
            root_mail = "root@%s" % _get_maindomain()
            for user in question.choices.keys():
                if root_mail in user_info(user)["mail-aliases"]:
                    question.default = user
                    break
```

Comment (alexAubin) [reply src/yunohost/app.py:2685]: Yolo let's see, thanks for the fix !

File: src/yunohost/app.py (hunk 2 above)
Comment (Psycojoker) [reply src/yunohost/app.py:2572]: Ah indeed, quite a lot of other PR has modified this code, sorry for not moving fast enough :< 

(don't hesitate to push on it ™)

**Comment by alexAubin** [comment]
Tests are passing ! Yolomergin !

**Pull request merged**